File: app/src/main/java/be/digitalia/fosdem/api/FosdemApi.java
package be.digitalia.fosdem.api;

import android.content.Context;
import android.os.AsyncTask;
import android.text.format.DateUtils;

import androidx.annotation.MainThread;
import androidx.annotation.NonNull;
import androidx.annotation.WorkerThread;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Transformations;

import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicBoolean;

import be.digitalia.fosdem.db.AppDatabase;
import be.digitalia.fosdem.db.ScheduleDao;
import be.digitalia.fosdem.livedata.LiveDataFactory;
import be.digitalia.fosdem.livedata.SingleEvent;
import be.digitalia.fosdem.model.Day;
import be.digitalia.fosdem.model.DetailedEvent;
import be.digitalia.fosdem.model.DownloadScheduleResult;
import be.digitalia.fosdem.model.RoomStatus;
import be.digitalia.fosdem.parsers.EventsParser;
import be.digitalia.fosdem.utils.network.HttpUtils;
import okio.BufferedSource;

/**
 * Main API entry point.
 *
 * @author Christophe Beyls
 */
public class FosdemApi {

	// 8:30 (local time)
	private static final long DAY_START_TIME = 8 * DateUtils.HOUR_IN_MILLIS + 30 * DateUtils.MINUTE_IN_MILLIS;
	// 19:00 (local time)
	private static final long DAY_END_TIME = 19 * DateUtils.HOUR_IN_MILLIS;

	private static final AtomicBoolean isLoading = new AtomicBoolean();
	private static final MutableLiveData<Integer> progress = new MutableLiveData<>();
	private static final MutableLiveData<SingleEvent<DownloadScheduleResult>> result = new MutableLiveData<>();
	private static LiveData<Map<String, RoomStatus>> roomStatuses;

	/**
	 * Download & store the schedule to the database.
	 * Only one thread at a time will perform the actual action, the other ones will return immediately.
	 * The result will be sent back in the consumable Result LiveData.
	 */
	@MainThread
	public static void downloadSchedule(@NonNull Context context) {
		if (!isLoading.compareAndSet(false, true)) {
			// If a download is already in progress, return immediately
			return;
		}
		final Context appContext = context.getApplicationContext();
		AsyncTask.THREAD_POOL_EXECUTOR.execute(() -> {
			downloadScheduleInternal(appContext);
			isLoading.set(false);
		});
	}

	@WorkerThread
	private static void downloadScheduleInternal(@NonNull Context context) {
		progress.postValue(-1);
		DownloadScheduleResult res = DownloadScheduleResult.error();
		try {
			ScheduleDao scheduleDao = AppDatabase.getInstance(context).getScheduleDao();
			HttpUtils.Response httpResponse = HttpUtils.get(
					FosdemUrls.getSchedule(),
					scheduleDao.getLastModifiedTag(),
					progress::postValue);
			if (httpResponse.source == null) {
				// Nothing to parse, the result is up-to-date.
				res = DownloadScheduleResult.upToDate();
				return;
			}

			try (BufferedSource source = httpResponse.source) {
				Iterable<DetailedEvent> events = new EventsParser().parse(source);
				int count = scheduleDao.storeSchedule(events, httpResponse.lastModified);
				res = DownloadScheduleResult.success(count);
			}

		} catch (Exception e) {
			e.printStackTrace();
			res = DownloadScheduleResult.error();
		} finally {
			progress.postValue(100);
			result.postValue(new SingleEvent<>(res));
		}
	}

	/**
	 * @return The current schedule download progress:
	 * -1   : in progress, indeterminate
	 * 0..99: progress value
	 * 100  : download complete or inactive
	 */
	public static LiveData<Integer> getDownloadScheduleProgress() {
		return progress;
	}

	public static LiveData<SingleEvent<DownloadScheduleResult>> getDownloadScheduleResult() {
		return result;
	}

	@MainThread
	public static LiveData<Map<String, RoomStatus>> getRoomStatuses(@NonNull Context context) {
		if (roomStatuses == null) {
			// The room statuses will only be loaded when the event is live.
			// Use the days from the database to determine it.
			final LiveData<List<Day>> daysLiveData = AppDatabase.getInstance(context).getScheduleDao().getDays();
			final LiveData<Boolean> scheduler = Transformations.switchMap(daysLiveData, days -> {
				final long[] startEndTimestamps = new long[days.size() * 2];
				int index = 0;
				for (Day day : days) {
					final long dayStart = day.getDate().getTime();
					startEndTimestamps[index++] = dayStart + DAY_START_TIME;
					startEndTimestamps[index++] = dayStart + DAY_END_TIME;
				}
				return LiveDataFactory.scheduler(startEndTimestamps);
			});
			final LiveData<Map<String, RoomStatus>> liveRoomStatuses = new LiveRoomStatusesLiveData();
			final LiveData<Map<String, RoomStatus>> offlineRoomStatuses = new MutableLiveData<>(Collections.emptyMap());
			roomStatuses = Transformations.switchMap(scheduler, isLive -> isLive ? liveRoomStatuses : offlineRoomStatuses);
			// Implementors: replace the above code with the next line to disable room status support
			// roomStatuses = new MutableLiveData<>();
		}
		return roomStatuses;
	}
}

File: app/src/main/java/be/digitalia/fosdem/api/FosdemApi.kt
package be.digitalia.fosdem.api

import android.util.Log
import androidx.annotation.MainThread
import be.digitalia.fosdem.alarms.AppAlarmManager
import be.digitalia.fosdem.api.network.HttpClient
import be.digitalia.fosdem.db.ScheduleDao
import be.digitalia.fosdem.flow.schedulerFlow
import be.digitalia.fosdem.flow.stateFlow
import be.digitalia.fosdem.model.DownloadScheduleResult
import be.digitalia.fosdem.model.LoadingState
import be.digitalia.fosdem.model.RoomStatus
import be.digitalia.fosdem.parsers.RoomStatusesParser
import be.digitalia.fosdem.parsers.ScheduleParser
import be.digitalia.fosdem.utils.BackgroundWorkScope
import be.digitalia.fosdem.utils.ByteCountSource
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.flatMapLatest
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOf
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import okio.buffer
import javax.inject.Inject
import javax.inject.Provider
import javax.inject.Singleton
import kotlin.math.pow
import kotlin.time.Duration
import kotlin.time.Duration.Companion.minutes
import kotlin.time.Duration.Companion.seconds
import kotlin.time.TimeMark
import kotlin.time.TimeSource

/**
 * Main API entry point.
 *
 * @author Christophe Beyls
 */
@Singleton
class FosdemApi @Inject constructor(
    private val httpClient: HttpClient,
    private val scheduleParserProvider: Provider<ScheduleParser>,
    private val scheduleDao: ScheduleDao,
    private val alarmManager: AppAlarmManager,
    private val timeSource: TimeSource
) {
    private var downloadJob: Job? = null
    private val _downloadScheduleState =
        MutableStateFlow<LoadingState<DownloadScheduleResult>>(LoadingState.Idle())

    /**
     * Download & store the schedule to the database.
     * Only a single Job will be active at a time.
     * The result will be notified through downloadScheduleState StateFlow.
     */
    @MainThread
    fun downloadSchedule(): Job {
        // Returns the download job in progress, if any
        return downloadJob ?: BackgroundWorkScope.launch {
            downloadScheduleInternal()
            downloadJob = null
        }.also {
            downloadJob = it
        }
    }

    private suspend fun downloadScheduleInternal() {
        _downloadScheduleState.value = LoadingState.Loading()
        val res = try {
            when (val response = httpClient.get(FosdemUrls.schedule, scheduleDao.lastModifiedTag.first())) {
                is HttpClient.Response.NotModified -> DownloadScheduleResult.UpToDate    // Nothing parsed, the result is up-to-date
                is HttpClient.Response.Success -> {
                    val result = withContext(Dispatchers.IO) {
                        val httpResponse = response.body
                        val length = httpResponse.contentLength
                        val source = if (length > 0L) {
                            // Broadcast the progression in percents, with a precision of 1/10 of the total file size
                            ByteCountSource(httpResponse.body, length / 10L) { byteCount ->
                                // Cap percent to 100
                                val percent = (byteCount * 100L / length).toInt().coerceAtMost(100)
                                _downloadScheduleState.value = LoadingState.Loading(percent)
                            }.buffer()
                        } else {
                            httpResponse.body
                        }

                        source.use {
                            val schedule = scheduleParserProvider.get().parse(source)
                            scheduleDao.storeSchedule(schedule, httpResponse.lastModified)
                        }
                    }
                    alarmManager.onScheduleRefreshed()
                    DownloadScheduleResult.Success(result)
                }
            }
        } catch (e: CancellationException) {
            throw e
        } catch (e: Exception) {
            Log.e(TAG, "Error while attempting to download the new schedule", e)
            DownloadScheduleResult.Error
        }
        _downloadScheduleState.value = LoadingState.Idle(res)
    }

    val downloadScheduleState: StateFlow<LoadingState<DownloadScheduleResult>> =
        _downloadScheduleState.asStateFlow()

    fun downloadScheduleResultConsumed() {
        _downloadScheduleState.update { state ->
            if (state is LoadingState.Idle) LoadingState.Idle() else state
        }
    }

    @OptIn(ExperimentalCoroutinesApi::class)
    val roomStatuses: Flow<Map<String, RoomStatus>> by lazy(LazyThreadSafetyMode.NONE) {
        stateFlow(BackgroundWorkScope, emptyMap()) {
            // The room statuses will only be loaded when the event is live.
            // Use the days from the database to determine it.
            val scheduler = scheduleDao.days.flatMapLatest { days ->
                val startEndTimestamps = LongArray(days.size * 2)
                var index = 0
                for (day in days) {
                    startEndTimestamps[index++] = day.startTime.toEpochMilli()
                    startEndTimestamps[index++] = day.endTime.toEpochMilli()
                }
                schedulerFlow(*startEndTimestamps)
                    .flowWhileShared(SharingStarted.WhileSubscribed())
            }
            scheduler.distinctUntilChanged().flatMapLatest { isLive ->
                if (isLive) {
                    buildLiveRoomStatusesFlow()
                        .flowWhileShared(SharingStarted.WhileSubscribed(5000L))
                }
                else flowOf(emptyMap())
            }
        }
        // Implementors: replace the above code block with the next line to disable room status support
        // emptyFlow()
    }

    /**
     * Builds a stateful cold Flow which loads and refreshes the Room statuses during the event.
     */
    private fun buildLiveRoomStatusesFlow(): Flow<Map<String, RoomStatus>> {
        var nextRefreshTime: TimeMark? = null
        var expirationTime: TimeMark? = null
        var retryAttempt = 0

        return flow {
            var nextRefreshDelay = nextRefreshTime?.let { -it.elapsedNow() } ?: Duration.ZERO

            if (expirationTime?.hasPassedNow() == true) {
                // When the data expires, replace it with an empty value
                emit(emptyMap())
            }

            while (true) {
                delay(nextRefreshDelay)

                var now: TimeMark
                nextRefreshDelay = try {
                    val response = httpClient.get(FosdemUrls.rooms) { httpResponse ->
                        httpResponse.body.use {
                            RoomStatusesParser().parse(it)
                        }
                    }
                    now = timeSource.markNow()

                    retryAttempt = 0
                    expirationTime = now + ROOM_STATUS_EXPIRATION_DELAY
                    emit(response.body)
                    ROOM_STATUS_REFRESH_DELAY
                } catch (e: CancellationException) {
                    throw e
                } catch (_: Exception) {
                    now = timeSource.markNow()

                    if (expirationTime?.hasPassedNow() == true) {
                        emit(emptyMap())
                    }

                    // Use exponential backoff for retries
                    val multiplier = 2.0.pow(retryAttempt)
                    retryAttempt++
                    (ROOM_STATUS_FIRST_RETRY_DELAY * multiplier).let {
                        // Avoid using minOf() or coerceAtMost() which cause boxing of inline class Duration
                        if (it > ROOM_STATUS_REFRESH_DELAY) ROOM_STATUS_REFRESH_DELAY else it
                    }
                }

                nextRefreshTime = now + nextRefreshDelay
            }
        }
    }

    companion object {
        private const val TAG = "FosdemApi"
        private val ROOM_STATUS_REFRESH_DELAY = 90.seconds
        private val ROOM_STATUS_FIRST_RETRY_DELAY = 30.seconds
        private val ROOM_STATUS_EXPIRATION_DELAY = 6.minutes
    }
}